src/nvim/eval.c

```diff
@@ -23985,50 +23985,23 @@ typval_T eval_call_provider(char *provider, char *method, list_T *arguments)
   return rettv;
 }
 
-bool eval_has_provider(const char *name)
-{
-#define CHECK_PROVIDER(name) \
-  if (has_##name == -1) { \
-    has_##name = !!find_func((char_u *)"provider#" #name "#Call"); \
-    if (!has_##name) { \
-      script_autoload("provider#" #name "#Call", \
-                      sizeof("provider#" #name "#Call") - 1, \
-                      false); \
-      has_##name = !!find_func((char_u *)"provider#" #name "#Call"); \
-    } \
-  }
-
-  static int has_clipboard = -1;
-  static int has_python = -1;
-  static int has_python3 = -1;
-  static int has_ruby = -1;
-
-  if (strequal(name, "clipboard")) {
-    CHECK_PROVIDER(clipboard);
-    return has_clipboard;
-  } else if (strequal(name, "python3")) {
-    CHECK_PROVIDER(python3);
-    return has_python3;
-  } else if (strequal(name, "python")) {
-    CHECK_PROVIDER(python);
-    return has_python;
-  } else if (strequal(name, "ruby")) {
-    bool need_check_ruby = (has_ruby == -1);
-    CHECK_PROVIDER(ruby);
-    if (need_check_ruby && has_ruby == 1) {
-      char *rubyhost = call_func_retstr("provider#ruby#Detect", 0, NULL, true);
-      if (rubyhost) {
-        if (*rubyhost == NUL) {
-          // Invalid rubyhost executable. Gem is probably not installed.
-          has_ruby = 0;
-        }
-        xfree(rubyhost);
-      }
-    }
-    return has_ruby;
+/// Check if a named provider exists.
+bool eval_has_provider(const char *provider)
+{
+  char varname[256];
+  int varname_len = snprintf(varname, sizeof(varname),
+                             "provider#%s#enabled", provider);
+
+  typval_T tv;
+  if (get_var_tv(varname, varname_len, &tv, NULL, false, false) == FAIL) {
+    return false;
```

Comment:
The old caching mechanism was to avoid searching all of the autoload locations over and over in case the symbol is never defined. 

Now if `provider#%s#enabled` is not defined after searching autoload, we should probably throw an error, except if 'loadplugins' (aka `p_lpl`, aka `--noplugin`) is false. Because all providers should always define this.

Reply:
has_provider() is called by has() so `has('amiga')` would cause an error. Likewise for all kinds of startup checks.

Reply:
Autoload would be triggered in any case, so we can check that `provider#x#Call` exists first.

Reply:
> The old caching mechanism was to avoid searching all of the autoload locations over and over in case the symbol is never defined.
> 
> Now if `provider#%s#enabled` is not defined after searching autoload, we should probably throw an error, except if 'loadplugins' (aka `p_lpl`, aka `--noplugin`) is false. Because all providers should always define this.

Sorry,  `:help autoload` suggests that I was wrong:

```
This also works when reading a variable that has not been set yet:

        :let l = foo#bar#lvar

However, when the autoload script was already loaded it won't be loaded again
for an unknown variable.
```

Reply:
> has('amiga')

Somewhat related:  it seems that `eval_has_provider()` is expected to trigger autoload. So after this PR, `has('amiga')` will always search for `provider#amiga` on the runtimepath, every time it is called. 

Not sure how to avoid that. Maybe if there's an internal table of known plugin/package names, then anything not in the "provider" key could be skipped by `eval_has_provider()` ...

Reply:
I think we should keep a whitelist with the known provider names. It seems like an anti-feature to me that user code can turn any `has("xyz")` to true. We can easily add more names to the list as we add more provider features.

Reply:
Ok, let's do that for now.